### ISSUE TITLE
use_enum_values dangerous from a typing perspective

### ISSUE THREAD

**Issue by tlambert03** [pydantic/pydantic#6565]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

Continuing the discussion from https://github.com/pydantic/pydantic/issues/6540

`use_enum_values` appears to be one of the only cases in pydantic where the type of the actual value in the model is not contained in the type annotation or the field info.  Consider the example code below... `Foo.enum`, despite being type annotated as a `MyEnum`, will be an `int`. 

If you ran mypy on that script, you would get:
```
error: Unsupported operand types for + ("MyEnum" and "int")
```

While the mypy plugin could potentially (and probably should) catch and fix that, it is not the only issue.  `Foo.model_fields` is *also* lying about the type it contains.  That seems to me to stand in as a pretty stark contrast to the rest of pydantic (as a parsing library) that pretty much guarantees that what's inside an instance of a model is consistent with the type annotations.

Are there other cases where you would have to not only look at the field annotation but also in the model config to check what the type of a given field might be?

### Example Code

```Python
from enum import Enum
from pydantic import BaseModel, ConfigDict

class MyEnum(Enum):
    A = 1
    B = 2

class Foo(BaseModel):
    enum: MyEnum
    model_config = ConfigDict(use_enum_values=True)

foo = Foo(enum=MyEnum.A)
val = foo.enum + 6
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0.2
        pydantic-core version: 2.1.2 release build profile
                 install path: /Users/talley/mambaforge/envs/test/lib/python3.11/site-packages/pydantic
               python version: 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:41) [Clang 15.0.7 ]
                     platform: macOS-13.4.1-arm64-arm-64bit
     optional deps. installed: ['typing-extensions']
```

**Comment by dmontagu** [comment]
`use_enum_values` is definitely an issue from a typing perspective, and I would strongly advise against its use. It was introduced in PR #127 over five years ago, at a time when a lot of this stuff was less mature, and I think if we had a chance to do it over it wouldn't be included.

For backwards compatibility reasons, I don't think it's a good idea for us to _change_ the behavior with that setting enabled, but I think it will probably make sense to deprecate/remove that config option at some point.

> That seems to me to stand in as a pretty stark contrast to the rest of pydantic (as a parsing library) that pretty much guarantees that what's inside an instance of a model is consistent with the type annotations.

Agreed

> Are there other cases where you would have to not only look at the field annotation but also in the model config to check what the type of a given field might be?

Not to my knowledge, and in general I think we'd like to reduce/resolve them if possible. Again, I would view the use of `use_enum_values = True` as an anti-pattern.

If you want to propose related changes to the library, I'm open to proposals.

**Comment by tlambert03** [comment]
> For backwards compatibility reasons, I don't think it's a good idea for us to change the behavior with that setting enabled, but I think it will probably make sense to deprecate/remove that config option at some point.

yep, makes sense.  certainly can't just break it unceremoniously :)  (v2 could be as good of a time to deprecate as any though?  given all the other breaking changes?)

> If you want to propose related changes to the library, docs, etc., I think we're open to proposals.

I can think of a couple things:

- you could emit a warning or even an exception if someone annotated a model with an enum, while also using `use_enum_values`


    ```python
    # Error
    class Model(BaseModel):
        x: MyEnum
        model_config = {'use_enum_values': True}
    
    # No Error
    class Model(BaseModel):
        x: Literal['value1', 'value2']
        model_config = {'use_enum_values': True}
    ```
    (of course, then you might want to associate the field with a specific enum, which I think should happen in `Field()`)


- I think for me, as a developer who sometimes consumes _other_ libraries models, the nicest thing would be to update `FieldInfo` in some way to indicate that you won't be getting what `model_config['field'].annotation` tells you.  I recognize that it's weird to "lie" there too and say something that wasn't annotated, but then again, perhaps it's better to be consistent with the rest of pydantic and show what the field value is actually going to be.  So,
    ```python
    class Model(BaseModel):
        x: MyEnum
        model_config = {'use_enum_values': True}

    print(Model.model_fields['x'].annotation)  # Annotated[Literal[...], MyEnum]  ... or something like that
    ```
    ... or anything else you can think of that wouldn't require someone to always check both the `FieldInfo` and the `ConfigDict` in order to determine what they're likely to find for that given field?

**Comment by ariebovenberg** [comment]
Just wanted to chime in to say that this tripped us up as well. 

Since V3 seems to be happening eventually, can this option be officially scheduled for removal then? Perhaps in combination with a deprecation warning introduced in a V2.x release?

**Comment by sydney-runkle** [comment]
@ariebovenberg,

> Since V3 seems to be happening eventually, can this option be officially scheduled for removal then? Perhaps in combination with a deprecation warning introduced in a V2.x release?

This should certainly be revisited V3, but we don't think it makes sense to introduce a deprecation warning for this field in a minor release, so a change like that will have to wait for V3. We'll keep everyone posted on this thread with regards to any changes for this setting!

**Comment by elpablete** [comment]
I would say that this is pretty much a case use for DeprecationWarning
